gluster_exporter/metric_volume.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	volumeLabels = []string{
+		"volume",
+	}
+
+	glusterVolumeTotalCount = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "gluster",
+			Name:      "volume_total_count",
+			Help:      "Total no of volumes",
+		},
+		[]string{},
+	)
+	glusterVolumeStartedCount = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "gluster",
+			Name:      "volume_started_count",
+			Help:      "Total no of started volumes",
+		},
+		[]string{},
+	)
+	glusterVolumeStoppedCount = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "gluster",
+			Name:      "volume_stopped_count",
+			Help:      "Total no of stopped volumes",
+		},
+		[]string{},
+	)
+	glusterVolumeBrickCount = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "gluster",
+			Name:      "volume_brick_count",
+			Help:      "Total no of bricks in volume",
+		},
+		volumeLabels,
+	)
+)
+
+func getVolumeLabels(volname string) prometheus.Labels {
+	return prometheus.Labels{
+		"volume": volname,
+	}
+}
+
+func VolumeCounts() {
+	// Load and parse volume list
+	fileObj, err := os.Open(getVolInfoFile())
+	if err != nil {
+		// TODO: log the error
+		return
+	}
+	defer fileObj.Close()
+	data, _ := ioutil.ReadAll(fileObj)
+	var volumes []GlusterVolume
+	if err := json.Unmarshal(data, &volumes); err != nil {
+		return
+	}
+
+	volCount, volStartCount, volStopCount := 0, 0, 0
+
+	volCount = len(volumes)
+	for _, volume := range volumes {
+		switch state := volume.State; state {
+		case "Started":
+			volStartCount += 1
+		case "Stopped":
+			volStopCount +=1
+		default:
+			// Volume is just created, nothing to do
+		}
+		volBrickCount := 0
+		for _, subvol := range volume.SubVolumes {
+			volBrickCount += len(subvol.Bricks)
+		}
+		glusterVolumeBrickCount.With(getVolumeLabels(volume.Name)).Set(float64(volBrickCount))
+	}
+	glusterVolumeTotalCount.With(prometheus.Labels{}).Set(float64(volCount))
+	glusterVolumeStartedCount.With(prometheus.Labels{}).Set(float64(volStartCount))
+	glusterVolumeStoppedCount.With(prometheus.Labels{}).Set(float64(volStopCount))
+}
+
+func init() {
+	prometheus.MustRegister(glusterVolumeTotalCount)
+	prometheus.MustRegister(glusterVolumeStartedCount)
+	prometheus.MustRegister(glusterVolumeStoppedCount)
+	prometheus.MustRegister(glusterVolumeBrickCount)
+
+
+	// Register to update this every 2 seconds
+	// Name, Callback Func, Interval Seconds
+	registerMetric("gluster_volume", VolumeCounts, 5)
```

Comment:
Will this be available in every node?

Reply:
@sabose thats a valid question. It should be run from only one node.
Will add the logic for the same.

Reply:
@aravindavk I understand you had some thought around this, to identify a leader node and from there only these kind of collectors should be run. Should we make a REST call to gd2 to find the peers and take the first entry as leader and if matches with current node then run this collector?

I hope also at configuration level we should be able to configure collectors which says if it should run from all the storage nodes or from leader only.

@aravindavk please suggest/comment.